src/pybind/mgr/rgw/module.py

```diff
@@ -146,18 +191,84 @@ def _cmd_rgw_realm_rm_zone_creds(self,
 
         return HandleCommandResult(retval=retval, stdout=out, stderr=err)
 
+    @CLICommand('rgw realm tokens', perm='r')
+    def list_realm_tokens(self):
+        realms_info = []
+        for realm_info in RGWAM(self.env).get_realms_info():
+            if not realm_info['master_zone_id']:
+                realms_info.append({'realm': realm_info['realm_name'], 'token': 'realm has no master zone'})
+            elif not realm_info['endpoint']:
+                realms_info.append({'realm': realm_info['realm_name'], 'token': 'master zone has no endpoint'})
+            elif not (realm_info['access_key'] and realm_info['secret']):
+                realms_info.append({'realm': realm_info['realm_name'], 'token': 'master zone has no access/secret keys'})
+            else:
+                keys = ['realm_name', 'realm_id', 'is_primary', 'endpoint', 'access_key', 'secret']
+                realm_token = RealmToken(**{k: realm_info[k] for k in keys})
+                realm_token_b = realm_token.to_json().encode('utf-8')
+                realm_token_s = base64.b64encode(realm_token_b).decode('utf-8')
+                realms_info.append({'realm': realm_info['realm_name'], 'token': realm_token_s})
+
+        return HandleCommandResult(retval=0, stdout=json.dumps(realms_info), stderr='')
+
+    @CLICommand('rgw zone update', perm='rw')
+    def update_zone_info(self, realm_name: str, zonegroup_name: str, zone_name: str, realm_token: str, endpoints: List[str]):
```

Comment:
where are the `endpoints` coming from here? Is it from the zone spec as documented in `mgr/rgw.rst`?

Reply:
These endpoints are provided by cephadm automatically (and asynchronously). The call is made from `mgr/cephadm/server.py::_update_rgw_endpoints` which is called once the rgw daemons are created successfully by cephadm.